### PR TITLE
use resue package to reuse alterUser node

### PR DESCRIPTION
**Pull request by jensenojs**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #11663

## What this PR does / why we need it:

use reuse package to reuse memory for alterUser node

There are at least four places need to release stmt memory

- func doComQuery
- planCache evict
- PrepareStmt.Close
- after func getRewriteSQLStmt

### CODE REVIEW

**Comment by matrix-meow** [comment]
@jensenojs Thanks for your contributions!

**Here are review comments for file pkg/frontend/computation_wrapper.go:**
Title: use resue package to reuse alterUser node

Body: This pull request aims to use the "reuse" package to reuse memory for the alterUser node. There are four places where statement memory needs to be released: func doComQuery, planCache evict, PrepareStmt.Close, and after func getRewriteSQLStmt. The changes made in the file pkg/frontend/computation_wrapper.go include adding a check for the stmt variable and calling the Free() method on it if it is not nil.

Review:

1. The title of the pull request is clear and concise, indicating the purpose of the changes made.

2. The body of the pull request provides a brief explanation of the changes and mentions the issue number that this pull request fixes.

3. The changes made in the file pkg/frontend/computation_wrapper.go are focused on memory management and freeing resources.

4. The addition of the check for the stmt variable and calling the Free() method if it is not nil is a good approach to ensure proper memory management.

Suggestions:

1. It would be helpful to provide more context or explanation in the body of the pull request about why reusing memory for the alterUser node is important and how it improves the performance or efficiency of the code.

2. Consider adding comments in the code to explain the purpose of the changes and provide clarity for future developers.

3. It would be beneficial to include test cases to ensure that the memory reuse is working correctly and does not introduce any regressions.

4. Consider optimizing the code further by identifying any other areas where memory can be reused or resources can be freed.

Security Issues:

No security issues were identified in the provided pull request.

**Here are review comments for file pkg/frontend/plan_cache.go:**
Title: Use resue package to reuse alterUser node

Body: This pull request aims to use the "reuse" package to reuse memory for the alterUser node. There are four places where statement memory needs to be released: `doComQuery`, `planCache evict`, `PrepareStmt.Close`, and after `getRewriteSQLStmt` function. The changes made in the `plan_cache.go` file include adding code to release statement memory when it cannot be cached, removing the unnecessary return statement, and adding code to release statement memory when the cache capacity is exceeded.

Changes in `pkg/frontend/plan_cache.go`:
- Added import statement for the "reuse" package.
- Added code to release statement memory when it cannot be cached.
- Removed unnecessary return statement.
- Added code to release statement memory when the cache capacity is exceeded.
- Added code to release statement memory when cleaning the cache.

Overall, the changes made in this pull request seem to address the issue of releasing statement memory in various scenarios. However, there are a few problems that need addressing:

1. In the `cache` function, when a plan is nil, the code currently releases memory for all statements in the `stmts` slice. This may lead to memory leaks if some of the statements are not nil. It would be better to only release memory for the specific statement that is nil, rather than releasing memory for all statements.

2. In the `cache` function, the `return` statement after releasing memory for the first nil plan will cause the function to exit prematurely. This means that the code to add the plan to the cache and remove the oldest plan when the cache capacity is exceeded will not be executed. The `return` statement should be removed to ensure that the code continues to execute.

3. In the `clean` function, the code currently iterates over the `lruList` and releases memory for all statements. However, the loop condition `i < pc.lruList.Len()` is incorrect because the length of the list changes as elements are removed. It would be better to use a `for range` loop to iterate over the list and release memory for each statement.

To optimize the changes made in this pull request, the following suggestions can be considered:

1. Instead of releasing memory for all statements in the `stmts` slice when a plan is nil, only release memory for the specific statement that is nil. This can be done by adding a check for nil before calling the `Free` method on the statement.

2. Remove the unnecessary `return` statement in the `cache` function to ensure that the code to add the plan to the cache and remove the oldest plan is executed.

3. Use a `for range` loop instead of a `for` loop with a fixed length in the `clean` function to iterate over the `lruList` and release memory for each statement.

By addressing these problems and considering the optimization suggestions, the codebase can be improved in terms of memory management and code readability.

**Here are review comments for file pkg/frontend/session.go:**
Title: use resue package to reuse alterUser node

Body: This pull request aims to use the "reuse" package to reuse memory for the alterUser node. There are four places in the code where statement memory needs to be released: `doComQuery`, `planCache evict`, `PrepareStmt.Close`, and after the `getRewriteSQLStmt` function. The changes made in the `pkg/frontend/session.go` file involve adding a defer statement to release the memory for each statement in the `statements` slice.

Review:

1. The title of the pull request is clear and concise, indicating the purpose of the changes.

2. The body of the pull request provides a brief explanation of what the changes aim to achieve and mentions the specific places where memory needs to be released.

3. The changes made in the `pkg/frontend/session.go` file involve adding a defer statement to release the memory for each statement in the `statements` slice. This is a good approach to ensure that memory is properly released.

Suggestions:

1. It would be helpful to provide more context or explanation about the "reuse" package and how it is being used to reuse memory for the alterUser node. This would help reviewers understand the impact and benefits of the changes.

2. It would be beneficial to include more information about the security implications of the changes. Are there any potential security vulnerabilities that need to be addressed? If so, it would be important to mention them and provide suggestions for mitigation.

3. Consider adding comments to the code to explain the purpose of the defer statement and the reason for releasing the memory for each statement in the `statements` slice.

Optimizations:

1. The changes made in the pull request seem to be focused on memory management and ensuring that memory is properly released. It would be helpful to provide some performance benchmarks or measurements to demonstrate the impact of these changes on memory usage and overall performance.

2. Consider conducting a thorough code review to identify any other areas where memory may need to be released or optimized. This could help further improve memory management and performance.

Overall, the changes made in the pull request seem to be addressing an important issue related to memory management. However, providing more context, addressing security concerns, and conducting a thorough code review could further enhance the quality and effectiveness of the changes.

**Here are review comments for file pkg/frontend/types.go:**
Title: use resue package to reuse alterUser node

Body: This pull request aims to use the "reuse" package to reuse memory for the alterUser node. There are four places in the code that need to release statement memory: func doComQuery, planCache evict, PrepareStmt.Close, and after func getRewriteSQLStmt. The changes made in the file pkg/frontend/types.go include adding a Free() method to the InternalCmdFieldList struct and calling the Free() method in the Close() method of the PrepareStmt struct.

Review:

1. In the InternalCmdFieldList struct, the Free() method is added, but it is empty. It is unclear what the purpose of this method is and whether it should have any functionality. It is recommended to either remove the Free() method or provide a clear explanation of its purpose and add the necessary functionality.

2. In the PrepareStmt struct, the Close() method is modified to call the Free() method of the PrepareStmt field. However, it is not clear what the PrepareStmt field represents and whether it is necessary to call the Free() method on it. It is recommended to provide more context and explanation for this change.

3. The changes made in the file pkg/frontend/types.go seem unrelated to the stated goal of reusing memory for the alterUser node. It is unclear how these changes contribute to the overall objective. It is recommended to provide more information and context for these changes.

4. It is not clear how the reuse package is being used to achieve the goal of reusing memory for the alterUser node. The pull request does not provide any details or code changes related to the reuse package. It is recommended to provide more information on how the reuse package is being utilized and how it helps in reusing memory.

Suggestions:

1. If the Free() method in the InternalCmdFieldList struct is not necessary, it should be removed to avoid confusion.

2. Provide more context and explanation for the changes made in the PrepareStmt struct. Clarify the purpose of the PrepareStmt field and why it is necessary to call the Free() method on it.

3. Provide more information and context for the changes made in the file pkg/frontend/types.go. Explain how these changes contribute to the goal of reusing memory for the alterUser node.

4. Provide more details on how the reuse package is being used and how it helps in reusing memory. Include code changes related to the reuse package to demonstrate its usage.

Optimizations:

It is difficult to suggest optimizations without a clear understanding of the purpose and functionality of the changes made in the pull request. Once the above issues are addressed and more information is provided, it may be possible to identify areas for optimization.

**Here are review comments for file pkg/sql/compile/sql_executor.go:**
Title: use resue package to reuse alterUser node

Body: This pull request aims to use the "reuse" package to reuse memory for the alterUser node. There are four places in the code that need to release statement memory: func doComQuery, planCache evict, PrepareStmt.Close, and after func getRewriteSQLStmt.

Changes in pkg/sql/compile/sql_executor.go:
- Added a defer statement to release the memory for each statement in the stmts slice.

Review:

1. The title of the pull request is clear and concise, indicating the purpose of the changes.

2. The body of the pull request provides a brief explanation of what the changes aim to achieve and mentions the specific places in the code that need to release statement memory.

3. The changes in sql_executor.go add a defer statement to release the memory for each statement in the stmts slice. This is a good improvement as it ensures that the memory is properly released even if an error occurs during parsing.

Suggestions:

1. It would be helpful to provide more context or explanation for why the memory needs to be released in those specific places. This would help reviewers understand the impact and importance of the changes.

2. Consider adding comments to the code to explain the purpose of the defer statement and the reason for releasing the memory in those specific places. This would make the code more self-explanatory and easier to maintain in the future.

3. It would be beneficial to run tests to ensure that the changes do not introduce any regressions or issues. This would provide more confidence in the correctness of the code changes.

Overall, the changes made in this pull request are a good improvement as they ensure proper memory management. However, providing more context and adding comments would enhance the clarity and maintainability of the code. Additionally, running tests would help ensure the correctness of the changes.

**Here are review comments for file pkg/sql/parsers/dialect/mysql/mysql_sql.go:**
Based on the provided information, the pull request aims to use the "reuse" package to reuse memory for the "alterUser" node in the MySQL SQL dialect parser. The changes made in the pull request are not provided, so it is not possible to analyze the specific code modifications. However, based on the description, there are four places where memory needs to be released:

1. `doComQuery` function
2. `planCache` eviction
3. `PrepareStmt.Close` function
4. After the `getRewriteSQLStmt` function

Without the actual code changes, it is difficult to provide specific feedback or identify any security issues. However, here are some general suggestions for optimizing the changes:

1. Ensure that the memory release is handled correctly in each of the mentioned places. This includes properly deallocating any resources and avoiding memory leaks.
2. Consider using appropriate error handling mechanisms to handle any potential errors during memory release.
3. Perform thorough testing to ensure that the memory reuse does not introduce any bugs or regressions in the codebase.
4. Document the changes made and provide clear comments explaining the purpose and functionality of the reused memory.

It is important to review the actual code changes in order to provide more specific and actionable feedback.

**Here are review comments for file pkg/sql/parsers/dialect/mysql/mysql_sql.y:**
Review:

Title: use resue package to reuse alterUser node

Body: This pull request aims to use the "reuse" package to reuse memory for the alterUser node. There are four places where statement memory needs to be released: `doComQuery`, `planCache evict`, `PrepareStmt.Close`, and after the `getRewriteSQLStmt` function. The changes made in the `mysql_sql.y` file involve modifying the `alter_user_stmt` section to use the `tree.NewAlterUser` function instead of directly assigning values to the `AlterUser` struct.

Overall, the changes made in this pull request seem to be focused on memory management and code organization. However, there are a few issues that need to be addressed:

1. Security Issue: The pull request does not mention any security-related changes, but it is important to ensure that the reuse of memory does not introduce any security vulnerabilities. It would be beneficial to conduct a thorough security review of the code changes to identify and address any potential security issues.

2. Lack of Explanation: The pull request does not provide a detailed explanation of why the reuse of memory is necessary or how it improves the codebase. It would be helpful to include more information about the specific benefits and optimizations achieved through this change.

3. Inconsistent Commenting: The changes made in the `mysql_sql.y` file include commenting out the previous assignment to the `AlterUser` struct and adding a new line of code using the `tree.NewAlterUser` function. This inconsistency in commenting style can make the code harder to understand and maintain. It would be better to either remove the commented-out code or provide a clear explanation for why it is being kept.

To optimize the changes made in this pull request, consider the following suggestions:

1. Provide a Detailed Explanation: Include a more detailed explanation in the pull request body about why the reuse of memory is necessary and how it improves the codebase. This will help reviewers and future developers understand the purpose and benefits of the changes.

2. Address Security Concerns: Conduct a thorough security review of the code changes to ensure that the reuse of memory does not introduce any security vulnerabilities. This may involve analyzing potential risks and implementing appropriate safeguards.

3. Consistent Commenting: Ensure that the commenting style is consistent throughout the codebase. Remove the commented-out code if it is no longer needed or provide a clear explanation for why it is being kept.

By addressing these issues and considering the suggested optimizations, the quality and maintainability of the codebase can be improved.

**Here are review comments for file pkg/sql/parsers/tree/alter.go:**
Review:

Title: use resue package to reuse alterUser node

Body: 
- This pull request aims to use the `reuse` package to reuse memory for the `alterUser` node.
- There are at least four places where statement memory needs to be released.
- The changes are made in the `alter.go` file in the `pkg/sql/parsers/tree` package.

Changes Made:
- Added import statement for the `reuse` package.
- Added an `init` function to create a pool for the `AlterUser` struct using the `reuse` package.
- Added a `Free` method to free the memory of the `AlterUser` struct using the `reuse` package.
- Modified the `NewAlterUser` function to use the `reuse` package to allocate memory for the `AlterUser` struct.

Feedback:
- The use of the `reuse` package to reuse memory for the `AlterUser` node is a good improvement.
- However, there are a few issues that need to be addressed:

1. In the `init` function, the `CreatePool` function is called with the `AlterUser` struct as the argument. However, the `AlterUser` struct does not implement the `reuse.Reusable` interface. To fix this, the `AlterUser` struct should implement the `reuse.Reusable` interface by adding the `Name` and `Free` methods.

2. In the `NewAlterUser` function, the `reuse.Alloc` function is called with `nil` as the argument. This will cause a panic at runtime. The `reuse.Alloc` function should be called with a valid `*AlterUser` pointer as the argument.

3. The `NewAlterUser` function has an additional parameter `commentOrAttribute` which is not mentioned in the pull request description. It is unclear why this parameter is added and how it should be used. The pull request description should provide more information about this change.

Optimization:
- The changes made in the pull request seem to be focused on memory management and do not have much scope for optimization. However, it is always a good practice to ensure that memory is properly released and reused to avoid memory leaks and improve performance.

Overall, the use of the `reuse` package to reuse memory for the `AlterUser` node is a good improvement. However, the issues mentioned above need to be addressed before merging the pull request.

**Here are review comments for file pkg/sql/parsers/tree/explain.go:**
Review:

Title: use resue package to reuse alterUser node

Body:
- The PR is an improvement.
- It fixes issue #11663.
- The purpose of the PR is to use the "reuse" package to reuse memory for the alterUser node.
- There are four places where memory needs to be released: `doComQuery`, `planCache evict`, `PrepareStmt.Close`, and `getRewriteSQLStmt`.
- The changes are made in the `explain.go` file.

Changes in `pkg/sql/parsers/tree/explain.go`:
- Added a new method `Free()` to the `explainImpl` struct.
- The `Free()` method does not have any implementation.

Feedback:
- The title of the pull request is clear and concise.
- The body provides relevant information about the purpose of the PR and the places where memory needs to be released.
- The changes in the `explain.go` file are minimal, adding a new method `Free()` to the `explainImpl` struct.
- However, the implementation of the `Free()` method is missing.

Suggestions:
- Implement the `Free()` method in the `explainImpl` struct to release the memory.
- Provide a clear explanation of how the memory is being reused and how the `Free()` method helps in releasing the memory.
- Consider adding test cases to ensure the correctness of the memory reuse and release.
- Consider providing more context and explanation in the pull request body to help reviewers understand the changes and their impact.

Security Issues:
- No security issues were identified in the provided information.

**Here are review comments for file pkg/sql/parsers/tree/prepare.go:**
Review:

Title: use resue package to reuse alterUser node

The title of the pull request is clear and concise. It indicates that the purpose of the pull request is to use the "resue" package to reuse the alterUser node.

Body:

The body of the pull request provides some additional context and information about the changes being made. It mentions that there are at least four places where statement memory needs to be released. It also references issue #11663, indicating that this pull request is related to fixing that issue.

Changes:

The changes made in the pull request are focused on the file "pkg/sql/parsers/tree/prepare.go". The only change made in this file is the addition of a new method called "Free" to the "prepareImpl" struct. This method does not have any implementation code.

Feedback:

1. Security: There are no apparent security issues in the changes made in this pull request.

2. Code Optimization: 

- The addition of the "Free" method to the "prepareImpl" struct seems to be incomplete. It does not have any implementation code, so it is not clear how it will be used or what it is intended to do. It would be helpful to provide more information or complete the implementation of this method.

- It is not clear how the "resue" package is being used to reuse the alterUser node. It would be helpful to provide more information or code changes related to this.

Suggestions:

1. Complete the implementation of the "Free" method in the "prepareImpl" struct. Provide the necessary code to release the statement memory in the four mentioned places.

2. Provide more information or code changes related to how the "resue" package is being used to reuse the alterUser node. This will help reviewers understand the changes better.

Overall, the pull request seems to be addressing an improvement and fixing an issue related to statement memory release. However, more information and code changes are needed to fully understand and review the changes made.

**Here are review comments for file pkg/sql/parsers/tree/show.go:**
Review:

Title: use resue package to reuse alterUser node

The title of the pull request is clear and concise. It indicates that the purpose of the pull request is to use the "reuse" package to reuse the alterUser node.

Body:

The body of the pull request provides some additional context and information about the changes made. It mentions that there are at least four places where memory needs to be released for the alterUser node. It also references issue #11663, indicating that this pull request is related to fixing that issue.

Changes:

The changes made in the pull request are focused on the file "pkg/sql/parsers/tree/show.go". The only change made is the addition of a new method called "Free" to the "showImpl" struct. This method does not contain any code.

Suggestions:

1. Security: It is important to ensure that memory is properly released to prevent memory leaks and potential security vulnerabilities. It would be helpful to provide more details about how the "reuse" package is being used to address this issue and ensure that memory is properly released.

2. Optimization: The addition of the "Free" method does not contain any code. It is unclear what the purpose of this method is and how it relates to the reuse of the alterUser node. It would be helpful to provide more information or remove this method if it is not necessary.

3. Documentation: It would be beneficial to provide more detailed documentation or comments explaining the purpose and usage of the "Free" method. This would help other developers understand the code and its intended functionality.

Overall, the pull request addresses the issue of memory release for the alterUser node. However, more information and clarification are needed to fully understand the changes and their impact.

**Here are review comments for file pkg/sql/parsers/tree/stmt.go:**
Review:

Title: use resue package to reuse alterUser node

Body: 
- This pull request aims to use the "reuse" package to reuse memory for the alterUser node.
- There are four places in the code that need to release statement memory.
- The changes made in the `stmt.go` file include adding a `Free()` method to the `statementImpl` struct.

Review:

1. The title of the pull request is clear and concise, indicating the purpose of the changes.

2. The body of the pull request provides some context about the changes and mentions the specific issue being addressed.

3. The changes made in the `stmt.go` file include adding a `Free()` method to the `statementImpl` struct. This method is intended to release the statement memory.

Suggestions:

1. It would be helpful to provide more details about the specific issue being addressed and how the reuse package is being used to solve it. This would provide more context for reviewers and make it easier to understand the changes.

2. It is unclear why the `Free()` method is added to the `statementImpl` struct. It would be helpful to provide more information about why this method is necessary and how it is used.

3. It is important to ensure that the `Free()` method is implemented correctly and releases the statement memory properly. Reviewers should carefully review the implementation of this method to ensure it is correct and does not introduce any memory leaks or other issues.

4. It would be beneficial to include test cases to verify the correctness of the changes and ensure that they do not introduce any regressions.

5. It is important to consider any potential security implications of the changes. Reviewers should carefully review the changes to ensure that they do not introduce any security vulnerabilities or weaken the security of the codebase.

Optimizations:

1. It would be helpful to provide more information about the performance improvements expected from using the reuse package. This would help reviewers understand the potential impact of the changes and prioritize their review accordingly.

2. It may be beneficial to provide benchmarks or performance measurements to demonstrate the performance improvements achieved by using the reuse package. This would provide more concrete evidence of the benefits of the changes and help reviewers assess their impact.

Overall, the pull request addresses a specific issue and makes changes to the `stmt.go` file to use the reuse package for memory reuse. However, more information is needed to fully understand the changes and their impact. Reviewers should carefully review the implementation of the `Free()` method and consider any potential security implications. Additionally, test cases and performance measurements would be beneficial to ensure the correctness and effectiveness of the changes.

**Here are review comments for file pkg/sql/plan/build_alter_table.go:**
Title: use resue package to reuse alterUser node

Body: This pull request aims to use the "reuse" package to reuse memory for the alterUser node. There are four places in the code that need to release statement memory: func doComQuery, planCache evict, PrepareStmt.Close, and after func getRewriteSQLStmt. The changes made in the file pkg/sql/plan/build_alter_table.go involve adding a defer statement to free the statement memory after it is used.

Review:

1. The title of the pull request is clear and concise, indicating the purpose of the changes made.

2. The body of the pull request provides a brief explanation of the changes and mentions the specific places where memory needs to be released.

3. The changes made in the file pkg/sql/plan/build_alter_table.go involve adding a defer statement to free the statement memory after it is used. This is a good practice to ensure that memory is properly released and can help prevent memory leaks.

Suggestions:

1. It would be helpful to provide more context or explanation about why the reuse package is being used and how it improves the code or performance. This would help reviewers understand the motivation behind the changes and evaluate their effectiveness.

2. It is important to ensure that the defer statement is placed correctly and that it is freeing the correct memory. Review the code carefully to verify that the defer statement is being used in the appropriate location and that it is freeing the statement memory as intended.

3. Consider adding comments to the code to explain the purpose of the defer statement and its relationship to the reuse package. This can help future developers understand the code and make any necessary modifications or improvements.

4. It would be beneficial to conduct thorough testing to ensure that the changes made in this pull request do not introduce any regressions or issues. Test cases should cover all the places where memory needs to be released to ensure that the reuse package is working correctly.

Overall, the changes made in this pull request seem reasonable and can help improve memory management in the code. However, providing more context, adding comments, and conducting thorough testing can further enhance the quality and maintainability of the codebase.

**Here are review comments for file pkg/sql/plan/build_show.go:**
Review:

Title: use resue package to reuse alterUser node

Body: 
- This pull request aims to use the "reuse" package to reuse memory for the alterUser node.
- There are four places in the code that need to release statement memory.
- The changes are made in the file `pkg/sql/plan/build_show.go`.

Changes in `pkg/sql/plan/build_show.go`:
- Line 1204: The `getRewriteSQLStmt` function is called with the `ctx` and `sql` parameters. The returned statement is assigned to the `stmt` variable.
- Line 1205: The `stmt` variable is replaced with `newStmt`.
- Line 1206: A `defer` statement is added to free the `newStmt` memory.
- Line 1211: The `stmt` variable is replaced with `newStmt`.
- Line 1212: The `stmt` variable is replaced with `newStmt`.
- Line 1220: The `stmt` variable is replaced with `newStmt`.
- Line 1221: A `defer` statement is added to free the `newStmt` memory.
- Line 1227: The `stmt` variable is replaced with `newStmt`.
- Line 1228: A `defer` statement is added to free the `newStmt` memory.

Overall, the changes in this pull request aim to improve memory management by using the `reuse` package to reuse memory for the alterUser node. The changes involve replacing the `stmt` variable with `newStmt` and adding `defer` statements to free the memory.

Suggestions:
1. It would be helpful to provide more context in the pull request body about why reusing memory for the alterUser node is important and how it improves performance or resource usage.
2. Consider adding comments in the code to explain the purpose of the changes and how they improve memory management.
3. Ensure that the `getRewriteSQLStmt` function is properly handling errors and returning appropriate error messages.
4. Consider running tests to verify that the changes do not introduce any regressions or issues.
5. Optimize the code by avoiding unnecessary calls to `getRewriteSQLStmt` if the same statement has already been obtained and reused.

Security Issues:
No security issues were identified in this pull request.

**Review by ouyuanning — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

**Review by iamlinjunhong — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Pull request merged**
